src/backend/distributed/planner/multi_physical_planner.c

```diff
@@ -1978,183 +1987,205 @@ BuildJobTreeTaskList(Job *jobTree)
+	 * Note that this has to be done on a copy, as the originals are needed
+	 * per target shard interval.
+	 */
+	foreach(restrictionCell, copiedRestrictionContext->relationRestrictionList)
+	{
+		RelationRestriction *restriction = lfirst(restrictionCell);
+		Index rteIndex = restriction->index;
+		List *originalBaserestrictInfo = restriction->relOptInfo->baserestrictinfo;
+
+		if (!allRelationsJoinedOnPartitionKey || allReferenceTables)
 		{
-			largestTableSize = tableSize;
-			anchorRangeTableId = tableId;
+			continue;
 		}
 
-		rangeTableIndex++;
-	}
+		shardOpExpressions = ShardIntervalOpExpressions(shardInterval, rteIndex);
 
-	/*
-	 * Ands are made implicit during shard pruning, as predicate comparison and
-	 * refutation depend on it being so. We need to make them explicit again so
-	 * that the query string is generated as (...) AND (...) as opposed to
-	 * (...), (...).
-	 */
-	whereClauseTree = (Node *) make_ands_explicit((List *) subquery->jointree->quals);
-	subquery->jointree->quals = whereClauseTree;
+		shardRestrictionList = make_simple_restrictinfo((Expr *) shardOpExpressions);
+		originalBaserestrictInfo = lappend(originalBaserestrictInfo,
+										   shardRestrictionList);
 
-	/* create tasks from every fragment combination */
-	foreach(fragmentCombinationCell, fragmentCombinationList)
-	{
-		List *fragmentCombination = (List *) lfirst(fragmentCombinationCell);
-		List *taskRangeTableList = NIL;
-		Query *taskQuery = copyObject(subquery);
-		Task *sqlTask = NULL;
-		StringInfo sqlQueryString = NULL;
+		restriction->relOptInfo->baserestrictinfo = originalBaserestrictInfo;
+	}
 
-		/* create tasks to fetch fragments required for the sql task */
-		List *uniqueFragmentList = UniqueFragmentList(fragmentCombination);
-		List *dataFetchTaskList = DataFetchTaskList(jobId, taskIdIndex,
-													uniqueFragmentList);
-		int32 dataFetchTaskCount = list_length(dataFetchTaskList);
-		taskIdIndex += dataFetchTaskCount;
 
-		ExtractRangeTableRelationWalker((Node *) taskQuery, &taskRangeTableList);
-		UpdateRangeTableAlias(taskRangeTableList, fragmentCombination);
+	/* mark that we don't want the router planner to generate dummy hosts/queries */
+	replacePrunedQueryWithDummy = false;
 
-		/* transform the updated task query to a SQL query string */
-		sqlQueryString = makeStringInfo();
-		pg_get_query_def(taskQuery, sqlQueryString);
+	/*
+	 * Use router select planner to decide on whether we can push down the query
+	 * or not. If we can, we also rely on the side-effects that all RTEs have been
+	 * updated to point to the relevant nodes and selectPlacementList is determined.
+	 */
+	routerPlannable = RouterSelectQuery(copiedQuery, copiedRestrictionContext,
+										&selectPlacementList, &selectAnchorShardId,
+										&relationShardList, replacePrunedQueryWithDummy);
 
-		sqlTask = CreateBasicTask(jobId, taskIdIndex, SQL_TASK, sqlQueryString->data);
-		sqlTask->dependedTaskList = dataFetchTaskList;
+	if (!routerPlannable)
```

Comment:
Hmm, I couldn't find such a path. But, I'm not enough confident to switch to an Assert. Are you OK with keeping it as an `if`?